src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -640,6 +701,56 @@ impl<'tcx> Tree {
+    /// Retags can violate the invariant on `latest_foreign_access`,
+    /// which states that a node's children survive accesses at least as strong as that node itself.
+    /// While retags act as a read for the offsets specified in the retag, it is also present at the other
+    /// offsets, but marked as "lazy." Such lazy nodes can, however, still be affected by foreign accesses.
+    /// This would violate the invariants that children always survive accesses at least as strong as the last
+    /// recorded foreign access at the parent.
+    /// To restore the invariant, we need to weaken the information stored in all parents, until it is weak enough
+    /// to encompass the newly added lazy node.
+    /// This function walks upwards from the newly inserted node, and ensures that all its parents are aware that the subtree
+    /// rooted at that parent might now survive fewer foreign accesses than it did before this node was inserted.
+    pub fn update_last_accessed_after_retag(
+        &mut self,
+        child_tag: BorTag,
+        new_perm: Permission,
+        prot: bool,
+    ) {
+        let mut current = self.tag_mapping.get(&child_tag).unwrap();
+        let strongest_survivable = new_perm.strongest_survivable_foreign_action(prot);
+        // We walk the tree upwards, until the invariant is restored
+        while let Some(next) = self.nodes.get(current).unwrap().parent {
+            current = next;
+            // record whether we did any change (if not, the invariant is restored).
+            let any_change = self.rperms.iter_mut_all().any(|(_, map)| {
+                match map.get_mut(current) {
+                    // if there is a permission, ensure that it's latest recorded foreign access is not stronger
+                    // than the `strongest_survivable` by the newly inserted child.
+                    // Returns `true` if it changed something, and in that case, we need to continue with the parent.
+                    Some(perm) => perm.reset_last_foreign_access_after_retag(strongest_survivable),
+                    // if there is no permission yet, it is still "default lazy".
+                    // in that case, we would need to compare with that node's original default permission.
+                    // But that depended on whether that node was protected when it was created, which is no
+                    // longer the case now.
+                    // Anyways, since the node is "default", its latest recorded foreign access will default to `None`,
+                    // but we might need to continue weakening the parent.
+                    // So we return `true` here, which is definitely correct, but maybe not fully optimal.
+                    // But in cases where we are imprecise, the iteration stops at the next initialized parent anyway,
+                    // so it is likely not a huge loss in practice.
+                    None => true,
```

Comment:
I'm still somewhat unhappy with this. In cases where a node is mostly default-initialized, this code here can become significant. A better solution might be to track the "default last foreign access" for not-yet-materialized nodes, similar to how we track the default permission for these. And then update that in here.

Reply:
Indeed, this seems to be the issue in `zip-equal` where a better handling of this makes things go 28% more brrr.